### PR TITLE
add API Gateway CustomAuthorizerEvent

### PR DESCRIPTION
**Pull request by keitakn**
Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `tsc` without errors.
- [x] Run `npm run lint package-name` if a `tslint.json` is present.

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `npm run new-package package-name`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

Document
- http://docs.aws.amazon.com/apigateway/latest/developerguide/use-custom-authorizer.html

### CODE REVIEW

File: types/aws-lambda/index.d.ts [hunk 1]
```diff
@@ -190,6 +197,7 @@ interface ProxyResult {
  */
 export type Handler = (event: any, context: Context, callback?: Callback) => void;
 export type ProxyHandler = (event: APIGatewayEvent, context: Context, callback?: ProxyCallback) => void;
+export type CustomAuthorizerHandler = (event: CustomAuthorizerEvent, context: Context, callback?: Callback) => void;
```
Comment (buggy) [inline types/aws-lambda/index.d.ts]: I think you should define a custom type for the callback parameter. That way you can check the structure passed to the callback is valid for a custom authorizer. See [Output from an Amazon API Gateway Custom Authorizer](http://docs.aws.amazon.com/apigateway/latest/developerguide/use-custom-authorizer.html) and how it's done for the ProxyHandler

Comment (keita-nishimoto) [inline types/aws-lambda/index.d.ts]: @buggy 

I think your opinion is correct.

I will make corrections as you have indicated.

Thank you.

Comment (keita-nishimoto) [inline types/aws-lambda/index.d.ts]: I did the addition of "CustomAuthorizerCallback" and "AuthResponse".

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15387/commits/60bda6cf4f8cd130f8a47a3dc3cdcd71bbffc6ea

File: types/aws-lambda/index.d.ts [hunk 2]
```diff
@@ -181,6 +188,45 @@ interface ProxyResult {
...
+    Action: string;
+    Effect: string;
+    Resource: string | [string];
+}
+
+/**
+ * API Gateway CustomAuthorizer AuthResponse.PolicyDocument.Statement.
+ * http://docs.aws.amazon.com/apigateway/latest/developerguide/use-custom-authorizer.html#api-gateway-custom-authorizer-output
+ */
+interface AuthResponseContext {
```
Comment (buggy) [inline types/aws-lambda/index.d.ts]: I think this is wrong. While the docs use the example `stringKey`, `numberKey` and `booleanKey` it seems like the context is actually a map and you can use any key name you like.

```typescript
interface AuthResponseContext {
  [name: string]: string | number | boolean;
}
```

File: types/aws-lambda/index.d.ts [hunk 3]
```diff
@@ -181,6 +188,45 @@ interface ProxyResult {
...
+}
+
+/**
+ * API Gateway CustomAuthorizer AuthResponse.PolicyDocument.Statement.
+ * http://docs.aws.amazon.com/apigateway/latest/developerguide/use-custom-authorizer.html#api-gateway-custom-authorizer-output
+ */
+interface Statement {
+    Action: string;
+    Effect: string;
+    Resource: string | [string];
```
Comment (buggy) [inline types/aws-lambda/index.d.ts]: Is this correct? The only examples I can find for Resource have a string. Is a string array valid?

File: types/aws-lambda/index.d.ts (hunk 2 above)
Comment (keita-nishimoto) [inline types/aws-lambda/index.d.ts]: My method was wrong.
I also thought you pointed out correctly.
I fixed the code.
Thank you.

File: types/aws-lambda/index.d.ts (hunk 3 above)
Comment (keita-nishimoto) [inline types/aws-lambda/index.d.ts]: @buggy 
Yes. In the document, only usage examples with string type are listed.

But, when I verified the operation with AWS, processing was completed normally even in the form of "[string]".

I understand this behavior as a specification for giving access to multiple resources with a single authorization process.

Comment (keita-nishimoto) [inline types/aws-lambda/index.d.ts]: I was taught the following documents from AWS engineers.
http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html

We can know that "Resources" is available in the array in this document.

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Permission",
      "Action": [
        "execute-api:Execution-operation"           
      ],
      "Resource": [
        "arn:aws:execute-api:region:account-id:api-id/stage/METHOD_HTTP_VERB/Resource-path"
      ]
    }
  ]
}
```

Comment (buggy) [inline types/aws-lambda/index.d.ts]: Thanks for that link. Then the last thing is that Action should also accept a string or an array of strings. After that I'm good to go.

Comment (keita-nishimoto) [inline types/aws-lambda/index.d.ts]: Thank you for a polite review.
I fixed a source code.

**Pull request merged**